### PR TITLE
refactor:support cose signature format

### PR DESCRIPTION
**Pull request by chloeyin**
## What
Refactor `notation-go` to support cose envelope
Background can be checked out in https://github.com/notaryproject/notation/discussions/278
Currently it uses dependencies from cose branch

## Major Changes
- Use package `github.com/notaryproject/notation-core-go/signature` to sign and verify
- Combine `runner` and `signer` to a `provider` for `pluginSigner` to sign and remove `pluginSigProvider`
- Add `builtinProvider` to support local signing and `externalProvider` to support signing by plugin
- Move `VerificationPlugin` and `VerificationPluginMinVersion` to `ExtendedAttributes`
- Retrieve `VerificationPlugin` and `VerificationPluginMinVersion` by calling `signerInfo.ExtendedAttribute`
- Add `SignatureMediaType` to `PutSignatureManifest` to support different types of signatures
- Support new [keySpec](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection)
- Add `GuessSignatureEnvelopeFormat` to inspect signature (This function may change later to better inspect a signature)
- Add sign/verify from file test cases
- Remove unused code
- Update go.mod dependency
- Some minor clean up

## Suggested review order

1. signature/plugin.go and signature/provider.go
2. signature/signer.go
3. signature/algorithm.go
4. All files except tests
5. signature/plugin_test.go and signature/provider_test.go
6. All test files left

Signed-off-by: zaihaoyin <zaihaoyin@microsoft.com>

### CODE REVIEW

File: signature/algorithm.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
+package signature
+
+import (
+	"errors"
+
+	"github.com/notaryproject/notation-core-go/signature"
+)
+
+// one of the following key spec name.
```
Comment (patrickzheng200) [inline signature/algorithm.go:9]: a bit confused by this comment. do you mean "supported key spec names"?

File: signature/algorithm.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+const (
+	RSA_2048 = "RSA-2048"
+	RSA_3072 = "RSA-3072"
+	RSA_4096 = "RSA-4096"
+	EC_256   = "EC-256"
+	EC_384   = "EC-384"
+	EC_521   = "EC-521"
+)
+
+// one of the following hash name.
```
Comment (patrickzheng200) [inline signature/algorithm.go:21]: same here. how about "supported hash names"

File: signature/algorithm.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+//
+// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection
+const (
+	SHA_256 = "SHA-256"
+	SHA_384 = "SHA-384"
+	SHA_512 = "SHA-512"
+)
+
+// one of the following signing algorithm name.
```
Comment (patrickzheng200) [inline signature/algorithm.go:30]: same here. "supported signing algorithm names"

File: signature/algorithm.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	switch k.Type {
+	case signature.KeyTypeEC:
+		switch k.Size {
+		case 256:
+			return EC_256
+		case 384:
+			return EC_384
+		case 521:
+			return EC_521
+		}
```
Comment (patrickzheng200) [inline signature/algorithm.go:56]: should we fail other k.Size values here?

File: signature/algorithm.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		switch k.Size {
+		case 2048:
+			return RSA_2048
+		case 3072:
+			return RSA_3072
+		case 4096:
+			return RSA_4096
+		}
+	}
+	return ""
```
Comment (patrickzheng200) [inline signature/algorithm.go:67]: should we fail other k.Type values?

File: signature/algorithm.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	switch k.Type {
+	case signature.KeyTypeEC:
+		switch k.Size {
+		case 256:
+			return SHA_256
+		case 384:
+			return SHA_384
+		case 521:
+			return SHA_512
+		}
```
Comment (patrickzheng200) [inline signature/algorithm.go:81]: same here

File: signature/algorithm.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		switch k.Size {
+		case 2048:
+			return SHA_256
+		case 3072:
+			return SHA_384
+		case 4096:
+			return SHA_512
+		}
+	}
+	return ""
```
Comment (patrickzheng200) [inline signature/algorithm.go:92]: same here

File: signature/algorithm.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	case signature.AlgorithmES512:
+		return ECDSA_SHA_512
+	case signature.AlgorithmPS256:
+		return RSASSA_PSS_SHA_256
+	case signature.AlgorithmPS384:
+		return RSASSA_PSS_SHA_384
+	case signature.AlgorithmPS512:
+		return RSASSA_PSS_SHA_512
+	}
+	return ""
```
Comment (patrickzheng200) [inline signature/algorithm.go:139]: should we fail other alg values?

File: signature/envelope_test.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+func init() {
+	msg := gcose.Sign1Message{
+		Headers:   gcose.NewSign1Message().Headers,
+		Payload:   []byte("valid"),
+		Signature: []byte("valid"),
+	}
+	validCoseSignatureEnvelope, _ = msg.MarshalCBOR()
+}
+
+// func setU
```
Comment (patrickzheng200) [inline signature/envelope_test.go:28]: nit: this line should be removed?

File: signature/provider.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+func (*builtinProvider) metadata() *plugin.Metadata {
+	// The only properties that are really relevant
+	// are the supported contract version and the capabilities.
+	// All other are just filled with meaningful data.
+	return &builtInPluginMetaData
+}
+
+// SetConfig set config when signing.
```
Comment (patrickzheng200) [inline signature/provider.go:54]: nit: SetConfig sets

File: signature/provider.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+func (*builtinProvider) SetConfig(map[string]string) {
+
+}
+
+// Run implement the plugin workflow.
```
Comment (patrickzheng200) [inline signature/provider.go:59]: nit: Run implements

File: signature/provider.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		return nil, fmt.Errorf("describe-key command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.DescribeKeyResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect describe-key response type '%T'", out)
+	}
+	return resp, nil
+}
+
+// Sign sign the digest by calling the underlying plugin.
```
Comment (patrickzheng200) [inline signature/provider.go:122]: nit: Sign signs

File: signature/provider.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// SetConfig set config when signing.
+func (*builtinProvider) SetConfig(map[string]string) {
+
+}
+
+// Run implement the plugin workflow.
+//
+// builtinProvider only support metadata and describe key.
```
Comment (patrickzheng200) [inline signature/provider.go:61]: nit: supports

File: plugin/plugin.go [hunk 14]
```diff
@@ -127,17 +125,21 @@ type DescribeKeyResponse struct {
 
 	// One of following supported key types:
 	// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection
-	KeySpec signer.KeySpec `json:"keySpec"`
+	KeySpec string `json:"keySpec"`
+
+	// Ordered list of certificates starting with leaf certificate
+	// and ending with root certificate.
```
Comment (patrickzheng200) [inline plugin/plugin.go:131]: looks like this comment should be removed?

File: signature/algorithm.go (hunk 4 above)
Comment (chloeyin) [reply signature/algorithm.go:56]: > should we fail other k.Size values here?

Do you mean to return an error instead of an empty string?

**Pull request closed without merging**